### PR TITLE
Enhance Cluster Health Check

### LINKED ISSUE #1821: test_nodereplacement_proactive failes with HEALTH_WARN 1 OSDs or CRUSH {nodes, device-classes} have {NOUP,NODOWN,NOIN,NOOUT} flags set
Ceph cluster health is not OK. Health: {health}"
            )
E           ocs_ci.ocs.exceptions.CephHealthException: Ceph cluster health is not OK. Health: HEALTH_WARN 1 OSDs or CRUSH {nodes, device-classes} have {NOUP,NODOWN,NOIN,NOOUT} flags set; Degraded data redundancy: 8285/204969 objects degraded (4.042%), 42 pgs degraded, 42 pgs undersized

### PR DESCRIPTION
**Pull request by prsurve**
fixes #1821 

Signed-off-by: Pratik Surve <pratiksurve2130@gmail.com>

### CODE REVIEW

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 1]
```diff
@@ -7,11 +7,12 @@
 from ocs_ci.framework.pytest_customization.marks import tier4a
 from ocs_ci.ocs.resources import pod
 from ocs_ci.framework.testlib import (
-    tier4, ManageTest, aws_platform_required, ignore_leftovers, ipi_deployment_required
-)
+    tier4, ManageTest, aws_platform_required, ignore_leftovers, ipi_deployment_required,
+    config)
```
Comment (RazTamir) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:11]: ) in new line

Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py]: Done

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 2]
```diff
@@ -97,4 +99,8 @@ def test_nodereplacement_proactive(self, pvc_factory, pod_factory, dc_pod_factor
         self.sanity_helpers.delete_resources()
         # Verify everything running fine
         log.info("Verifying All resources are Running and matches expected result")
-        self.sanity_helpers.health_check()
+        self.sanity_helpers.health_check(cluster_check=False)
```
Comment (PrasadDesala) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:102]: Here, you are just ignore the cluster health check. Still the ceph_health_check will happen and test case may fail if cluster is not healthy.

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 3]
```diff
@@ -97,4 +99,8 @@ def test_nodereplacement_proactive(self, pvc_factory, pod_factory, dc_pod_factor
...
+        log.info("Verifying ceph cluster status")
+        ceph_health_check(
+            namespace=config.ENV_DATA['cluster_namespace'], tries=10,
```
Comment (PrasadDesala) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:105]: You dont need to specify namespace here as by default the function takes ocs namespace

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py [hunk 4]
```diff
@@ -97,4 +104,29 @@ def test_nodereplacement_proactive(self, pvc_factory, pod_factory, dc_pod_factor
         self.sanity_helpers.delete_resources()
         # Verify everything running fine
         log.info("Verifying All resources are Running and matches expected result")
-        self.sanity_helpers.health_check()
+        all_pod_obj = pod.get_all_pods(
```
Comment (ebenahar) [inline tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:107]: self.sanity_helpers.health_check() calls CephCluster.cluster_health_check() which scans the cluster resources and checks their status. 
Why not using it? If something is missing from it, let's add it there

Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:107]: for sanity_helpers.health_check() timeout is 60 and first health_check() checks for ceph status and then for pod status. For this test case we need to make sure all pods are in running state and then check ceph status

Comment (ebenahar) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:107]: how about modifying sanity_helpers.health_check() so ceph_health_check() will be called after ceph_cluster.cluster_health_check()?

Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:107]: This can be done will work on changing health_check()

Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:107]: Updated function to check pod status first and then ceph_health

Comment (PrasadDesala) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:105]: > for sanity_helpers.health_check() timeout is 60 and first health_check() checks for ceph status and then for pod status. For this test case we need to make sure all pods are in running state and then check ceph status

Why? If mon or osd stays at pending state after replacement the ceph cluster health will not be healthy.

File: tests/sanity_helpers.py [hunk 5]
```diff
@@ -31,10 +31,11 @@ def health_check(self, cluster_check=True, tries=20):
         logger.info("Checking cluster and Ceph health")
         node.wait_for_nodes_status()
 
-        ceph_health_check(namespace=config.ENV_DATA['cluster_namespace'], tries=tries)
         if cluster_check:
             self.ceph_cluster.cluster_health_check(timeout=60)
 
+        ceph_health_check(namespace=config.ENV_DATA['cluster_namespace'], tries=tries)
```
Comment (PrasadDesala) [inline tests/sanity_helpers.py:37]: Why are we making this change? 
In cluster health we are checking ceph health status, the test case will fail at this phase if it didn't reach to healthy state in 60sec(which is default) which is quite impossible in tier4 cases.
So, IMO better to not make any changes to this file as you are increasing the tries attempts to 60 in test case.

Please clarify it I am wrong here.

File: tests/manage/z_cluster/nodes/test_node_replacement_proactive.py (hunk 4 above)
Comment (prsurve) [reply tests/manage/z_cluster/nodes/test_node_replacement_proactive.py:105]: if we do ceph_health_check first we won't get actual result of failure i.e pods reached to running  or not that's why moved health_check down

File: tests/sanity_helpers.py (hunk 5 above)
Comment (prsurve) [reply tests/sanity_helpers.py:37]: if we do ceph_health_check first there is no use of  ceph_cluster.cluster_health_check () because of ceph_health_check passes then all pods will be in running state

Comment (PrasadDesala) [reply tests/sanity_helpers.py:37]: > if we do ceph_health_check first there is no use of ceph_cluster.cluster_health_check () because of ceph_health_check passes then all pods will be in running state. I don't see any issue here moving ceph_health_check() down

Not always. I think if some non ceph related pod is not in running state (like drain canary pod or collector pod) still the ceph health will shows as healthy :)
Just putting my point. 
If others are okay I am fine with this change.

Comment (ebenahar) [reply tests/sanity_helpers.py:37]: we'll need to run tier4 suites for verification with this change

Comment (prsurve) [reply tests/sanity_helpers.py]: reverted back to default

Comment (prsurve) [reply tests/sanity_helpers.py]: updated cluster_health_check to check pod status and expected count

File: ocs_ci/ocs/cluster.py [hunk 6]
```diff
@@ -586,6 +693,230 @@ def get_throughput_percentage(self):
         logging.info(f"The throughput percentage of the cluster is {throughput_percentage}%")
         return throughput_percentage
 
+    def osd_health_check(self, count):
+        """
+        Osd health check based on pod count
```
Comment (ebenahar) [inline ocs_ci/ocs/cluster.py:698]: I'd change 'Osd' to 'OSD'. Also in other occurrences

File: ocs_ci/ocs/cluster.py [hunk 7]
```diff
@@ -586,6 +693,230 @@ def get_throughput_percentage(self):
...
+
+        Args:
+            count (int): Expected number of Osd pods
+
+        Raises:
+            OsdCountException: if osd pod count doesn't match
```
Comment (ebenahar) [inline ocs_ci/ocs/cluster.py:704]: add an empty line below. Please do in all functions

File: ocs_ci/ocs/cluster.py [hunk 8]
```diff
@@ -586,6 +693,230 @@ def get_throughput_percentage(self):
...
+        """
+        timeout = 10 * len(self.pods)
+        logger.info(f"Expected OSDs = {count}")
```
Comment (ebenahar) [inline ocs_ci/ocs/cluster.py:707]: Change to "Expected OSDs count: ..."

File: ocs_ci/ocs/cluster.py [hunk 9]
```diff
@@ -586,6 +693,230 @@ def get_throughput_percentage(self):
...
+        try:
+            assert self.POD.wait_for_resource(
+                condition='Running', selector=self.osd_selector,
+                resource_count=count, timeout=timeout, sleep=3,
+            )
+
+            actual_osds = len(pod.get_osd_pods())
+
+            assert count == actual_osds, f"Expected {count},  Got {actual_osds}"
```
Comment (ebenahar) [inline ocs_ci/ocs/cluster.py:716]: change to: "Expected OSD count {count}..."

File: ocs_ci/ocs/cluster.py [hunk 10]
```diff
@@ -586,6 +693,230 @@ def get_throughput_percentage(self):
...
+        Mgr health check based on pod count
+
+        Args:
+            count (int): Expected number of Mgr pods
+
+        Raises:
+            MgrCountException: if mgr pod count doesn't match
+        """
+        timeout = 10 * len(self.pods)
+        logger.info(f"Expected Mgr = {count}")
```
Comment (ebenahar) [inline ocs_ci/ocs/cluster.py:735]: same as above

**Comment by ebenahar** [comment]
@prsurve Thanks for adding these verifications!

The scope of this PR has changed. Please update the title accordingly

File: ocs_ci/ocs/cluster.py (hunk 8 above)
Comment (prsurve) [reply ocs_ci/ocs/cluster.py]: Done

File: ocs_ci/ocs/cluster.py (hunk 10 above)
Comment (prsurve) [reply ocs_ci/ocs/cluster.py]: Done

File: ocs_ci/ocs/cluster.py (hunk 7 above)
Comment (prsurve) [reply ocs_ci/ocs/cluster.py:703]: Done

File: ocs_ci/ocs/cluster.py (hunk 6 above)
Comment (prsurve) [reply ocs_ci/ocs/cluster.py]: Done

File: ocs_ci/ocs/cluster.py (hunk 9 above)
Comment (prsurve) [reply ocs_ci/ocs/cluster.py]: Done

**Comment by ebenahar** [comment]
@prsurve please rebase

**Comment by prsurve** [comment]
> @prsurve please rebase

Rebase Done

File: ocs_ci/ocs/cluster.py [hunk 11]
```diff
@@ -632,6 +738,238 @@ def time_taken_to_complete_rebalance(self, timeout=600):
...
+        """
+        timeout = 10 * len(self.pods)
+        logger.info(f"Expected Rgw count= {count}")
+        try:
+            assert self.POD.wait_for_resource(
+                condition='Running', selector=self.rgw_selector,
+                resource_count=count, timeout=timeout, sleep=3,
+            )
+
+            actual_rgw = 1
```
Comment (RazTamir) [inline ocs_ci/ocs/cluster.py:818]: that's not actual but expected right?

Comment (prsurve) [reply ocs_ci/ocs/cluster.py:818]: yes

Comment (prsurve) [reply ocs_ci/ocs/cluster.py]: Done

File: ocs_ci/ocs/cluster.py [hunk 12]
```diff
@@ -815,9 +815,9 @@ def rgw_health_check(self, count):
                 resource_count=count, timeout=timeout, sleep=3,
             )
 
-            actual_rgw = 1
+            expected_rgw = 1
```
Comment (RazTamir) [inline ocs_ci/ocs/cluster.py:818]: well,

If the expected is 1 and you will pass count=1 I'm not sure what this function does.
Don't we need to call a function to get the actual RGW pods?

Comment (prsurve) [reply ocs_ci/ocs/cluster.py]: I have update get_rgw_pod function to get number of rgw pod running in the cluster and return list object

**Comment by RazTamir** [comment]
Great PR

File: ocs_ci/ocs/resources/pod.py [hunk 13]
```diff
@@ -474,13 +474,18 @@ def get_rgw_pod(rgw_label=constants.RGW_APP_LABEL, namespace=None):
...
     Returns:
         Pod object: rgw pod object
     """
     namespace = namespace or config.ENV_DATA['cluster_namespace']
     rgws = get_pods_having_label(rgw_label, namespace)
-    rgw_pod = Pod(**rgws[0])
+    if not return_list:
+        rgw_pod = Pod(**rgws[0])
+    else:
+        rgw_pod = [Pod(**rgw) for rgw in rgws]
```
Comment (RazTamir) [inline ocs_ci/ocs/resources/pod.py:488]: rgw_pod = [Pod(**rgw) for rgw in rgws] if return_list else Pod(**rgws[0])

File: ocs_ci/ocs/resources/pod.py [hunk 14]
```diff
@@ -1259,3 +1264,60 @@ def wait_for_dc_app_pods_to_reach_running_state(dc_pod_obj):
...
+
+def get_rbdplugin_pods(rbdplugin_label=constants.CSI_RBDPLUGIN_LABEL, namespace=None):
+    """
+    Fetches info about rbdplugin pods in the cluster
+
+    Args:
+        rbdplugin_label (str): label associated with rbdplugin pods
+            (default: constants.CSI_RBDPLUGIN_LABEL)
+        namespace (str): Namespace in which ceph cluster lives
+            (default: defaults.ROOK_CLUSTER_NAMESPACE)
```
Comment (RazTamir) [inline ocs_ci/ocs/resources/pod.py:1277]: the real value is None

File: ocs_ci/ocs/resources/pod.py [hunk 15]
```diff
@@ -1259,3 +1264,60 @@ def wait_for_dc_app_pods_to_reach_running_state(dc_pod_obj):
...
+
+    Returns:
+        list : of cephfsplugin pod objects
+    """
+    namespace = namespace or config.ENV_DATA['cluster_namespace']
+    cephfsplugins = get_pods_having_label(cephfsplugin_label, namespace)
+    cephfsplugin_pods = [Pod(**cephfsplugin) for cephfsplugin in cephfsplugins]
+    return cephfsplugin_pods
```
Comment (RazTamir) [inline ocs_ci/ocs/resources/pod.py:1304]: please merge this with the rbdplugin function so you'll be able to use one function based on the plugin_label you pass

File: ocs_ci/ocs/resources/pod.py [hunk 16]
```diff
@@ -1259,3 +1264,60 @@ def wait_for_dc_app_pods_to_reach_running_state(dc_pod_obj):
...
+        namespace (str): Namespace in which ceph cluster lives
+            (default: defaults.ROOK_CLUSTER_NAMESPACE)
+
+    Returns:
+        list : of draincanary pod objects
+    """
+    namespace = namespace or config.ENV_DATA['cluster_namespace']
+    draincanarys = get_pods_having_label(draincanary_label, namespace)
+    draincanarys_pods = [Pod(**draincanary) for draincanary in draincanarys]
+    return draincanarys_pods
```
Comment (RazTamir) [inline ocs_ci/ocs/resources/pod.py:1323]: same

File: ocs_ci/ocs/resources/pod.py (hunk 15 above)
Comment (vasukulkarni) [reply ocs_ci/ocs/resources/pod.py:1304]: +1

File: ocs_ci/ocs/resources/pod.py (hunk 16 above)
Comment (vasukulkarni) [reply ocs_ci/ocs/resources/pod.py:1323]: also we have https://github.com/red-hat-storage/ocs-ci/blob/master/ocs_ci/ocs/resources/pod.py#L701

File: tests/e2e/scale/test_pvc_creation_deletion_scale.py [hunk 17]
```diff
@@ -109,9 +109,10 @@ def test_multiple_pvc_creation_deletion_scale(self, namespace, access_mode, inte
         # TODO: Update below code with google API, to record value in spreadsheet
         # TODO: For now observing Google API limit to write more than 100 writes
         log_path = f"{ocsci_log_path()}/{self.sc_obj}-{access_mode}"
-        csv_obj = csv.writer(open(f"{log_path}-creation-time.csv", "w"))
-        for k, v in pvc_create_time.items():
-            csv_obj.writerow([k, v])
+        with open(f"{log_path}-creation-time.csv", "w") as fd:
```
Comment (vasukulkarni) [inline tests/e2e/scale/test_pvc_creation_deletion_scale.py:112]: should be a separate PR if not already fixed.

File: tests/manage/storageclass/test_storageclass_invalid.py [hunk 18]
```diff
@@ -79,7 +79,8 @@ def invalid_storageclass(request):
     yaml_path = os.path.join(
         request.param['template_dir'], "storageclass.yaml"
     )
-    yaml_data = yaml.safe_load(open(yaml_path, 'r'))
+    with open(yaml_path, 'r') as fd:
```
Comment (vasukulkarni) [inline tests/manage/storageclass/test_storageclass_invalid.py:82]: separate PR please

**Comment by prsurve** [comment]
I guess I made something wrong while rebasing

**Comment by prsurve** [comment]
Closing and recreating new pr
